feat(image-select): add image select field to org unit form DHIS2-11175

Mohammer5 opened the pull request:
Closes DHIS2-11175
([link](https://jira.dhis2.org/browse/DHIS2-11175))

The image select works the following way:
* The image select has the following states: `loading`, `error`, `initialized`, `pending`, `hasInitialValue` & `removed`
  * `hasInitialValue` & `removed` are being used to display a text that says that the form has to be submitted in order to save the removal
  * `initialized` is being used to check whether the image is still `PENDING` or whether an error has occured (neither `PENDING` nor `STORED`)
  * `pending` causes a text to be rendered instead of an image, informing the user that the image is being processed on the server side
* When the user uploads an image, the post request to `fileResources` will always respond with `storageStatus: 'PENDING'`
  * So in order to display the image, the image select checks the `storageStatus` periodically (every second; up to 5 times) to update the `pending` status and display the image in case the status has changed. The text for the pending state just remains the same and is supposed to be meaningful even when the maximum amount of polls have been made and the status is still pending

#### OU has no image
![image](https://user-images.githubusercontent.com/10101831/123427116-7fd67700-d5c4-11eb-97be-badef3fd8686.png)

#### Image is being uploaded
![image](https://user-images.githubusercontent.com/10101831/123429032-c331e500-d5c6-11eb-9034-c94002b527cb.png)

#### Image is being processed
![image](https://user-images.githubusercontent.com/10101831/123427525-05f2bd80-d5c5-11eb-8009-76f5e405851c.png)

#### Image form value is dirty (image has been uploaded and is pending or stored or file has been removed)
![image](https://user-images.githubusercontent.com/10101831/123427626-291d6d00-d5c5-11eb-8be7-d9cc7500b01a.png)

#### error occurred while uploading
![image](https://user-images.githubusercontent.com/10101831/123428161-cd071880-d5c5-11eb-8edb-093d7d26cc12.png)

Birkbjo commented:
Code looks good to me!
However it would be great if we could streamline the UI a little bit. Doesn't necessarily need to be identical, but would be great to move it closer to eg. how we select icons:
![image](https://user-images.githubusercontent.com/13852438/123438001-8ff45380-d5d0-11eb-9e0b-b1427269e0e2.png)

eg. change buttons to be of the same color. And I don't think we really need a `reset`-button, as that is basically done by cancelling the form. I was at least a bit unsure about what it did at first.

Mohammer5 commented:
I didn't add the reset button either, but added it after I realised that cancelling the form will also discard all other potentially made changes. In my experience (which is very subjective of course) I always add pictures as the very last thing when editing profiles/forms and if you don't have the original on your computer, cancelling the whole form could be frustrating.
What do you think?

I'll adjust the button styles!

cooper-joe commented on src/i18n/i18n_module_en.properties:
```diff
@@ -2259,3 +2259,14 @@ open_periods_after_co_end_date=Open periods after category option end date
+org_unit_image_image_upload_error_text=There is an error with the image:
+org_unit_image_image_pending_text=The image has been uploaded and is being processed.
+org_unit_image_storage_status_error=Image could not be stored, storageStatus is:
+org_unit_image_save_reminder=Please submit the form to save your changes
```
I suggest `Submit the form to save changes`

cooper-joe commented on src/i18n/i18n_module_en.properties:
```diff
@@ -2259,3 +2259,14 @@ open_periods_after_co_end_date=Open periods after category option end date
+label_organisation_unit_image=Organisation unit image
+org_unit_image_alt_text=Organisation unit image
+org_unit_image_no_image_text=Organisation unit does not have an image
+org_unit_image_loading_image_data_text=Loading image data
```
I think we can remove `data` here and just use `Loading image`

cooper-joe commented on src/i18n/i18n_module_en.properties:
```diff
@@ -2259,3 +2259,14 @@ open_periods_after_co_end_date=Open periods after category option end date
+org_unit_image_loading_image_data_text=Loading image data
+org_unit_image_uploading_image_text=Uploading image
+org_unit_image_image_upload_error_text=There is an error with the image:
+org_unit_image_image_pending_text=The image has been uploaded and is being processed.
```
I suggest `Upload successful, processing image`

cooper-joe reviewed (commented):
I agree with @Birkbjo that it would be good to match the button styles with the existing Maintenance app.

Regarding the `Reset` button, I think it might be best to adjust the label to `Reset changes`. This makes it clear that it will reset back to the previous state, not to an empty state. I would also suggest hiding the `Reset changes` button until changes are made.

I've also made some suggestions to the helper texts.

Mohammer5 commented:
@Birkbjo I've changed a little bit of code, nothing wild
https://github.com/dhis2/maintenance-app/blob/c2417f9ed294652bf44a05e0c08e02a47b3a5e01/src/forms/form-fields/image-select.js#L331

---

@birk & @cooper-joe I've updated the styles, see below

#### OU has no image
![image](https://user-images.githubusercontent.com/10101831/123803454-99d9c780-d8ec-11eb-86d4-98453938eb11.png)

#### Image is being uploaded
![image](https://user-images.githubusercontent.com/10101831/123803541-acec9780-d8ec-11eb-9025-cd693761d52a.png)

#### Image is being processed

##### Replacing the old image
![image](https://user-images.githubusercontent.com/10101831/123805446-5d0ed000-d8ee-11eb-90e8-c856c757f90f.png)

##### New image
![image](https://user-images.githubusercontent.com/10101831/123805664-8e879b80-d8ee-11eb-91e7-9e9eb4222ad2.png)

#### Image form value is dirty (image has been uploaded and is pending or stored or file has been removed)

##### Org unit has no image so far
![image](https://user-images.githubusercontent.com/10101831/123803990-18cf0000-d8ed-11eb-8d26-d519c3727149.png)

##### Org unit had a different image
![image](https://user-images.githubusercontent.com/10101831/123804057-2c7a6680-d8ed-11eb-84b4-200b952ce9f8.png)

##### Previous image has been removed
![image](https://user-images.githubusercontent.com/10101831/123804113-3a2fec00-d8ed-11eb-9801-b1d59351d752.png)

#### error occurred while uploading
![image](https://user-images.githubusercontent.com/10101831/123807859-84ff3300-d8f0-11eb-9412-ed205f171308.png)

cooper-joe commented:
I think in the `OU has no image`, first example, it's not necessary to show the `x` icon - this could be misinterpreted as a button that has some function, or as an actual image of an `x`. I think not showing anything makes it clear there is no image.

...and just to check, is `Something went wrong  ¯\_(ツ)_/¯` actually being displayed? or is this an example error message from backend?

Mohammer5 commented:
@cooper-joe 

> I think in the OU has no image, first example, it's not necessary to show the x icon - this could be misinterpreted as a button that has some function, or as an actual image of an x. I think not showing anything makes it clear there is no image.

I assume same argument applies to when there's an error?

> ...and just to check, is Something went wrong ¯\_(ツ)_/¯ actually being displayed? or is this an example error message from backend?

Yes, that's just an example backend error :laughing:

cooper-joe commented:
> I assume same argument applies to when there's an error?

Yeah, that too - didn't notice that 👍

Mohammer5 commented:
@cooper-joe I've removed the "x" entirely, which changes 3 different states:

#### Error (This time with a real error message :laughing:):
![image](https://user-images.githubusercontent.com/10101831/123913443-82e3b580-d97e-11eb-96be-1848204d45ce.png)

#### No image:
![image](https://user-images.githubusercontent.com/10101831/123915551-ecfd5a00-d980-11eb-86d2-a8f9ebfaf5af.png)

#### Image has been removed:
![image](https://user-images.githubusercontent.com/10101831/123916073-8167bc80-d981-11eb-9ec0-ceeb59b2be4a.png)
